docs/upload_specifications.ADR.md

```diff
@@ -0,0 +1,170 @@
+
+## Narrative Bulk Upload cell operations
+
+When the bulk upload cell sees a `upload_specification` data type, it calls out to the StS
+endpoint detailed below to get the parsed specifications, and then autofills itself with
+those specifications.
+
+## Staging service upload specification endpoint
+
+The StS endpoint responds to an HTTP GET with a file list in a `files` URL parameter. It is
+extremely unlikely that there will be a large enough set of data types that URL length limits
+will be reached so this seems adequate. The StS will respond with the contents of the files
+parsed into a structure that is similar to that of the input spec, mapped by data type:
+
+```
+Headers:
+Authorization: <KBase token>
+
+GET /bulk_specification/?files=<file1.csv[,file2.csv,...]>
+
+Response:
+{"types": {
+    "gff_metagenome": [
+        {"fasta_file": "mygenome.fa",
+         "gff_file": "mygenome.gff3",
+         "genome_name": "mygenomeobject",
+         "source": null,
+         "release": null,
+         "genetic_code: 11,
+         "generate_missing_genes: 0
+         },
+        {"fasta_file": "mygenome2.fa",
+         "gff_file": "mygenome2.gff3",
+         "genome_name": "mygenomeobject2",
+         "source": "yermumspoo",
+         "release": 30456,
+         "genetic_code: 11,
+         "generate_missing_genes: 1
+         },
+         ...
+    ],
+    "sra_reads": [
+        ...
+    ]
+    ...
+}
+```
+
+The order of the input structures MUST be the same as the order in the input files.
+
+Notably, the service will provide the contents of the files as is and will not perform most error
+checking, including for missing or unknown input parameters. Most error checking will be performed
+in the bulk upload cell configuration tab like other uploads, allowing for a consistent user
+experience.
+
+### Missing files
+
+If the StS cannot find one or more files it will return a 404 with the body contents:
+```
+{"error": "Cannot_find_all_files",
+ "missing_files": [<list of missing filenames>]
+}
+```
+
+### Unparseable files
+
+If the StS cannot parse one or more of the files it will return a 400 with the body contents:
+```
+{"error": "Cannot_parse_all_files",
+ "unparseable_files": [{"filename": <filename>,
+                        "reason": <message>
+                        },
+                        ...
+                       ]
+}
+```
+The service will check that the data type is valid and that rows >=2 all have the same number of
+entries, but will not do further error checking.
+
+Note in this case the service MUST log the stack trace along with the filename for each invalid
+file.
+
+## Alternatives explored
+
+* We considered parsing the files in the Narrative backend (pulling the files as is from the StS)
+  but the current design allows other services and UIs reusing the parsing code. There doesn't
+  otherwise seem to be a strong reason to include the code one place or the other so we decided
+  to include it in the StS.
```

Comment:
Should we also note the pros and cons of adding it to a dynamic service or a standalone core service, or a BFF service somewhere?

Reply:
Hmm... I mean it seems pretty crazy to build another service to parse staging service files when the staging service is right there, so much so that I didn't even think of it, but if you think that's important I have no problem putting it in

Reply:
That's what we did [here](https://github.com/kbase/execution_engine2/blob/develop/docs/adrs/rework-batch-analysis-architecture.md#alternatives-considered) so mentioned it just in case

Reply:
In that case there was already external batch code so there was precedence. Here I don't think we have that situation, but again, if people think I should add it in I'm fine with that